firebase-firestore/src/main/java/com/google/firebase/firestore/model/mutation/FieldMask.java

```diff
@@ -27,13 +28,30 @@
  */
 public final class FieldMask {
   public static FieldMask fromSet(Set<FieldPath> mask) {
-    return new FieldMask(mask);
+    return new FieldMask(mask, Scope.BY_MASK);
+  }
+
+  public static FieldMask allFieldsMask() {
+    return new FieldMask(new HashSet<FieldPath>(), Scope.ALL_FIELDS);
+  }
+
+  public static FieldMask emptyMask() {
+    return new FieldMask(new HashSet<FieldPath>(), Scope.NONE);
```

Comment:
I am still trying to get it to work without this change (will continue tomorrow).

Reasons I advocate against it:
- It exposes a lot of state on the FieldMask class, some of which is mutually exclusive and only sensical if the consumer understand the underlying states. I wish we could somehow redesign this so that the mask type is hidden from the API.
- It breaks `getMask()`
- It breaks `covers()` (which can be fixed)

We would also need to fix equals+hashCode. 

I will report back tomorrow if I can come up with a solution.

Reply:
Follow up PR: https://github.com/firebase/firebase-android-sdk/pull/2983

Let me know what you think.

Note that the changes I made uncover a versioning issue that exists in this PR as well (there is a comment in the code changes explaining it).

Reply:
LGTM for the follow up, and thanks for doing it!

Re: version issue, this could be fixed by having a version update in the final mutation. We do need to serialize this as well. Thoughts?